theia-rust-docker/Dockerfile

```diff
@@ -0,0 +1,56 @@
+#Can't use node image, because it creates a user that conflicts
+FROM ubuntu:18.04
+
+RUN apt-get update -yq \
+    && apt-get install curl gnupg build-essential gcc g++ gdb make python sudo git -yq 
+
+#allow a user to become root in the container
+ADD sudoers /etc/sudoers
+
+ARG GITHUB_TOKEN
+ARG NODE_OPTIONS="--max-old-space-size=4096"
+ARG NODE_VERSION=10
+
+ADD install_node.sh /root/install_node.sh
+RUN bash /root/install_node.sh --version $NODE_VERSION
+ADD nvm_setup.sh /root/nvm_setup.sh
+
+#install theia
+ENV THEIA_RUST_APP_PATH /root/theia_rust_app
+RUN mkdir $THEIA_RUST_APP_PATH
+RUN . /root/nvm_setup.sh && cd $THEIA_RUST_APP_PATH && yarn global add node-gyp
+ADD next.package.json $THEIA_RUST_APP_PATH/package.json
+RUN git clone https://github.com/eclipse/theia-cpp-extension.git /root/theia-cpp-extension
+RUN cp -R /root/theia-cpp-extension/packages $THEIA_RUST_APP_PATH/.
+RUN . /root/nvm_setup.sh && cd $THEIA_RUST_APP_PATH && yarn --cache-folder $THEIA_RUST_APP_PATH/ycache 
+RUN . /root/nvm_setup.sh && cd $THEIA_RUST_APP_PATH/packages/cpp-debug && yarn --cache-folder $THEIA_RUST_APP_PATH/ycache 
+RUN . /root/nvm_setup.sh && cd $THEIA_RUST_APP_PATH && yarn theia build
+
+# when creating a container a volume is mapped to /root/rust
+# to allow to make persistent changes
+ARG rust_channel=nightly-2019-07-19
+
+ENV RUSTUP_HOME /root/rust/rustup
+ENV CARGO_HOME /root/rust/cargo
+
+RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs > /tmp/sh.rustup.rs && sh /tmp/sh.rustup.rs -y
+
+#install language server dependencies for theia
+RUN . /root/rust/cargo/env && rustup toolchain install $rust_channel
+RUN . /root/rust/cargo/env && rustup default $rust_channel
+RUN . /root/rust/cargo/env && rustup component add rust-src rust-docs rls rust-analysis rustfmt clippy
+
+RUN . /root/rust/cargo/env && cargo +$rust_channel install racer
```

Comment:
This step fails for me now with a compilation error, using the default rust channel setup a few lines up (i.e. building "manually", without using the provided makefile). 

```
Step 26/34 : RUN . /root/rust/cargo/env && cargo +$rust_channel install racer
 ---> Running in e7ce53e2092f
    Updating crates.io index
 Downloading crates ...
  Downloaded racer v2.1.24
  Installing racer v2.1.24
 Downloading crates ...
  Downloaded derive_more v0.13.0
  Downloaded bitflags v1.1.0
  Downloaded clap v2.33.0
  Downloaded lazycell v1.2.1
  Downloaded log v0.4.8
  Downloaded env_logger v0.6.2
  Downloaded racer-cargo-metadata v0.1.1
  Downloaded humantime v1.2.0
  Downloaded rls-span v0.5.2
  Downloaded lazy_static v1.3.0
  Downloaded textwrap v0.11.0
  Downloaded atty v0.2.13
  Downloaded regex v1.2.0
  Downloaded strsim v0.8.0
  Downloaded cfg-if v0.1.9
  Downloaded racer-interner v0.1.0
  Downloaded quote v0.6.13
  Downloaded unicode-width v0.1.5
  Downloaded termcolor v1.0.5
  Downloaded vec_map v0.8.1
  Downloaded serde v1.0.98
  Downloaded proc-macro2 v0.4.30
  Downloaded ansi_term v0.11.0
  Downloaded syn v0.15.42
  Downloaded serde_json v1.0.40
  Downloaded quick-error v1.2.2
  Downloaded thread_local v0.3.6
  Downloaded serde_derive v1.0.98
  Downloaded rustc-ap-syntax v542.0.0
  Downloaded utf8-ranges v1.0.3
  Downloaded libc v0.2.60
  Downloaded aho-corasick v0.7.4
  Downloaded unicode-xid v0.1.0
  Downloaded memchr v2.2.1
  Downloaded itoa v0.4.4
  Downloaded regex-syntax v0.6.10
  Downloaded rustc-ap-serialize v542.0.0
  Downloaded ryu v1.0.0
  Downloaded rustc-ap-rustc_macros v542.0.0
  Downloaded rustc-ap-syntax_pos v542.0.0
  Downloaded rustc-ap-rustc_target v542.0.0
  Downloaded smallvec v0.6.10
  Downloaded rustc-ap-rustc_data_structures v542.0.0
  Downloaded rustc-ap-rustc_errors v542.0.0
  Downloaded scoped-tls v1.0.0
  Downloaded rustc-ap-rustc_lexer v542.0.0
  Downloaded ucd-util v0.1.5
  Downloaded itertools v0.8.0
  Downloaded synstructure v0.10.2
  Downloaded indexmap v1.0.2
  Downloaded jobserver v0.1.16
  Downloaded rustc-rayon v0.2.0
  Downloaded rustc-ap-arena v542.0.0
  Downloaded rustc-ap-graphviz v542.0.0
  Downloaded crossbeam-utils v0.6.6
  Downloaded annotate-snippets v0.6.1
  Downloaded rustc-rayon-core v0.2.0
  Downloaded ena v0.13.0
  Downloaded rustc-hash v1.0.1
  Downloaded stable_deref_trait v1.1.1
  Downloaded crossbeam-deque v0.2.0
  Downloaded either v1.5.2
  Downloaded parking_lot v0.7.1
  Downloaded num_cpus v1.10.1
  Downloaded byteorder v1.3.2
  Downloaded crossbeam-epoch v0.3.1
  Downloaded crossbeam-utils v0.2.2
  Downloaded lock_api v0.1.5
  Downloaded parking_lot_core v0.4.0
  Downloaded arrayvec v0.4.11
  Downloaded memoffset v0.2.1
  Downloaded nodrop v0.1.13
  Downloaded scopeguard v0.3.3
  Downloaded owning_ref v0.4.0
  Downloaded rand v0.6.5
  Downloaded rustc_version v0.2.3
  Downloaded rand_chacha v0.1.1
  Downloaded rand_core v0.4.0
  Downloaded rand_xorshift v0.1.1
  Downloaded rand_os v0.1.3
  Downloaded rand_isaac v0.1.1
  Downloaded autocfg v0.1.5
  Downloaded rand_pcg v0.1.2
  Downloaded rand_hc v0.1.0
  Downloaded rand_core v0.3.1
  Downloaded rand_jitter v0.1.4
  Downloaded semver v0.9.0
  Downloaded semver-parser v0.7.0
   Compiling autocfg v0.1.5
   Compiling arrayvec v0.4.11
   Compiling semver-parser v0.7.0
   Compiling libc v0.2.60
   Compiling cfg-if v0.1.9
   Compiling rand_core v0.4.0
   Compiling nodrop v0.1.13
   Compiling lazy_static v1.3.0
   Compiling memoffset v0.2.1
   Compiling proc-macro2 v0.4.30
   Compiling scopeguard v0.3.3
   Compiling rustc-rayon-core v0.2.0
   Compiling log v0.4.8
   Compiling byteorder v1.3.2
   Compiling unicode-xid v0.1.0
   Compiling stable_deref_trait v1.1.1
   Compiling smallvec v0.6.10
   Compiling syn v0.15.42
   Compiling rustc-rayon v0.2.0
   Compiling indexmap v1.0.2
   Compiling either v1.5.2
   Compiling rustc-ap-graphviz v542.0.0
   Compiling memchr v2.2.1
   Compiling serde v1.0.98
   Compiling bitflags v1.1.0
   Compiling ryu v1.0.0
   Compiling ucd-util v0.1.5
   Compiling unicode-width v0.1.5
   Compiling scoped-tls v1.0.0
   Compiling rustc-ap-rustc_target v542.0.0
   Compiling annotate-snippets v0.6.1
   Compiling rustc-ap-syntax v542.0.0
   Compiling quick-error v1.2.2
   Compiling itoa v0.4.4
   Compiling termcolor v1.0.5
   Compiling utf8-ranges v1.0.3
   Compiling rustc-ap-rustc_lexer v542.0.0
   Compiling strsim v0.8.0
   Compiling vec_map v0.8.1
   Compiling ansi_term v0.11.0
error[E0658]: use of unstable library feature 'rustc_private': mainly needed for compiler internals
   --> /root/rust/cargo/registry/src/github.com-1ecc6299db9ec823/rustc-ap-rustc_lexer-542.0.0/src/lib.rs:689:33
    |
689 |             || (c > '\x7f' && c.is_xid_start())
    |                                 ^^^^^^^^^^^^
    |
    = note: for more information, see https://github.com/rust-lang/rust/issues/27812
    = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'rustc_private': mainly needed for compiler internals
   --> /root/rust/cargo/registry/src/github.com-1ecc6299db9ec823/rustc-ap-rustc_lexer-542.0.0/src/lib.rs:708:33
    |
708 |             || (c > '\x7f' && c.is_xid_continue())
    |                                 ^^^^^^^^^^^^^^^
    |
    = note: for more information, see https://github.com/rust-lang/rust/issues/27812
    = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.
error: Could not compile `rustc-ap-rustc_lexer`.
warning: build failed, waiting for other jobs to finish...
error: failed to compile `racer v2.1.24`, intermediate artifacts can be found at `/tmp/cargo-installcDtghX`

Caused by:
  build failed
The command '/bin/sh -c . /root/rust/cargo/env && cargo +$rust_channel install racer' returned a non-zero code: 101
```

Reply:
update: I can't build as well with the `stable` channel and also if I use the makefile. Maybe something was broken Dockerfile?

Reply:
It seems that there was an update in rust somewhere. The Racer component no longer compiles.  Strange thing is that this happens even while setting the rust channel to a specific date. Have to look into this.

Reply:
I just tried again, and now the image builds. Strange.

Reply:
Ah, maybe the build working again has to do with the commit you added.

Reply:
I temporarily disabled installing Racer.